File: parens.swift
#!/usr/bin/env swift

func _parens(_ prefix: String, _ left: Int, _ right: Int) {
    if left == 0 && right == 0 {
        print(prefix)
    }
    if left > 0 {
        _parens(prefix+"(", left-1, right)
    }
    if right > left {
        _parens(prefix+")", left, right-1)
    }
}

func parens(_ n: Int) {_parens("", n, n)}

parens(Int(CommandLine.arguments[1])!)

File: parens.rs
use std::env;

fn _parens(prefix: String, left: i32, right: i32) {
    if left == 0 && right == 0 {
        println!("{}", prefix);
    }

    if left > 0 {
        _parens(prefix.clone() + "(", left - 1, right);
    }

    if right > left {
        _parens(prefix.clone() + ")", left, right - 1);
    }
}

fn parens(n: i32) {
    _parens("".to_string(), n, n);
}

fn main() {
    let args: Vec<String> = env::args().collect();
    parens(args[1].parse::<i32>().unwrap());
}
